internal/provider/resource_credential.go

```diff
@@ -6,209 +6,307 @@ import (
+var _ resource.ResourceWithConfigure = &credentialResource{}
+var _ resource.ResourceWithImportState = &credentialResource{}
 
-		CreateContext: resourceCredentialCreate,
-		ReadContext:   resourceCredentialRead,
-		UpdateContext: resourceCredentialUpdate,
-		DeleteContext: resourceCredentialDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: resourceCredentialImporter,
-		},
+func NewCredentialResource() resource.Resource {
+	return &credentialResource{}
+}
 
-		Schema: map[string]*schema.Schema{
-			"cloud": {
+type credentialResource struct {
+	client *juju.Client
+}
+
+type credentialResourceModel struct {
+	Cloud                types.Object `tfsdk:"cloud"`
+	Attributes           types.Map    `tfsdk:"attributes"`
+	AuthType             types.String `tfsdk:"auth_type"`
+	ClientCredential     types.Bool   `tfsdk:"client_credential"`
+	ControllerCredential types.Bool   `tfsdk:"controller_credential"`
+	Name                 types.String `tfsdk:"name"`
+
+	// ID required by the testing framework
+	ID types.String `tfsdk:"id"`
+}
+
+func (c *credentialResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_credential"
+}
+
+func (c *credentialResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "A resource that represent a credential for a cloud.",
+		Blocks: map[string]schema.Block{
+			"cloud": schema.SingleNestedBlock{
 				Description: "JuJu Cloud where the credentials will be used to access",
-				Type:        schema.TypeList,
-				Optional:    true,
-				Computed:    true,
-				ForceNew:    true,
-				MaxItems:    1,
-				Elem: &schema.Resource{
-					Schema: map[string]*schema.Schema{
-						"name": {
-							Description: "The name of the cloud",
-							Type:        schema.TypeString,
-							Required:    true,
-						},
+				Attributes: map[string]schema.Attribute{
+					"name": schema.StringAttribute{
+						Description: "The name of the cloud",
+						Required:    true,
 					},
 				},
+				PlanModifiers: []planmodifier.Object{
+					objectplanmodifier.RequiresReplace(),
+				},
 			},
-			"attributes": {
+		},
+		Attributes: map[string]schema.Attribute{
+			"attributes": schema.MapAttribute{
 				Description: "Credential attributes accordingly to the cloud",
-				Type:        schema.TypeMap,
+				ElementType: types.StringType,
 				Optional:    true,
-				Elem:        &schema.Schema{Type: schema.TypeString, Optional: true},
 			},
-			"auth_type": {
+			"auth_type": schema.StringAttribute{
 				Description: "Credential authorization type",
-				Type:        schema.TypeString,
 				Required:    true,
 			},
-			"client_credential": {
+			"client_credential": schema.BoolAttribute{
 				Description: "Add credentials to the client",
-				Type:        schema.TypeBool,
 				Optional:    true,
-				Default:     false,
+				Computed:    true,
```

Comment:
I think I need to ponder about this a bit, the reason is the following error, but not sure if it's because of something that I'm  doing with that attribute in one of the CRUD methods:

```
Error while loading schemas for plugin components: Failed to obtain provider
        schema: Could not load the schema for provider
        registry.terraform.io/hashicorp/juju: failed to retrieve schema from provider
        "registry.terraform.io/hashicorp/juju": Schema Using Attribute Default For
        Non-Computed Attribute: Attribute "client_credential" must be computed when
        using default. This is an issue with the provider and should be reported to
        the provider developers...
```